python/tests/test_combinatorics.py

```diff
@@ -420,3 +433,331 @@ def test_is_symmetrical(self):
+        tables = tskit.TableCollection(sequence_length=1.0)
+
+        c1 = tables.nodes.add_row(flags=tskit.NODE_IS_SAMPLE, time=0)
+        c2 = tables.nodes.add_row(flags=tskit.NODE_IS_SAMPLE, time=0)
+        p = tables.nodes.add_row(flags=tskit.NODE_IS_SAMPLE, time=1)
+
+        tables.edges.add_row(left=0, right=1, parent=p, child=c1)
+        tables.edges.add_row(left=0, right=1, parent=p, child=c2)
+
+        self.verify_errors(tables.tree_sequence())
+
+    def verify_errors(self, ts):
+        with self.assertRaises(ValueError):
+            ts.first().count_topologies()
+        with self.assertRaises(ValueError):
+            list(ts.count_topologies())
+
+    def test_standard_msprime_migrations(self):
+        for num_populations in range(2, 5):
+            samples = [5] * num_populations
+            ts = self.simulate_multiple_populations(samples)
+            self.verify_topologies(ts)
+
+    def test_complex_dtwf_topologies(self):
+        migration_matrix = np.zeros((4, 4))
+        population_configurations = [
+            msprime.PopulationConfiguration(
+                sample_size=10, initial_size=10, growth_rate=0
+            ),
+            msprime.PopulationConfiguration(
+                sample_size=10, initial_size=10, growth_rate=0
+            ),
+            msprime.PopulationConfiguration(
+                sample_size=10, initial_size=10, growth_rate=0
+            ),
+            msprime.PopulationConfiguration(
+                sample_size=0, initial_size=10, growth_rate=0
+            ),
+        ]
+        demographic_events = [
+            msprime.PopulationParametersChange(population=1, time=0.1, initial_size=5),
+            msprime.PopulationParametersChange(population=0, time=0.2, initial_size=5),
+            msprime.MassMigration(time=1.1, source=0, dest=2),
+            msprime.MassMigration(time=1.2, source=1, dest=3),
+            msprime.MigrationRateChange(time=2.1, rate=0.3, matrix_index=(2, 3)),
+            msprime.MigrationRateChange(time=2.2, rate=0.3, matrix_index=(3, 2)),
+        ]
+        ts = msprime.simulate(
+            migration_matrix=migration_matrix,
+            population_configurations=population_configurations,
+            demographic_events=demographic_events,
+            random_seed=2,
+            model="dtwf",
+        )
+
+        self.verify_topologies(ts)
+
+    def simulate_multiple_populations(self, sample_sizes):
+        d = len(sample_sizes)
+        M = 0.2
+        m = M / (2 * (d - 1))
+
+        migration_matrix = [
+            [m if k < d and k == i + 1 else 0 for k in range(d)] for i in range(d)
+        ]
+
+        pop_configurations = [
+            msprime.PopulationConfiguration(sample_size=size) for size in sample_sizes
+        ]
+        return msprime.simulate(
+            population_configurations=pop_configurations,
+            migration_matrix=migration_matrix,
+            recombination_rate=0.1,
+        )
```

Comment:
As there's a lot going on in the PartialTopologyCounter, it'd be nice to see a few simple tests for it here. Nothing exhaustive, just a handful of simple examples demonstrating that it does what it's supposed to.

Reply:
Definitely, I think the tests definitely make it easier to pick apart.